cmd/snap-confine/udev-support.c

```diff
@@ -18,290 +18,355 @@
-	udev_s->tagname_len = sc_must_snprintf(udev_s->tagname, MAX_BUF,
-					       "%s", security_tag);
-	for (size_t i = 0; i < udev_s->tagname_len; i++)
-		if (udev_s->tagname[i] == '.')
-			udev_s->tagname[i] = '_';
+	/* Allow access to /dev/nvidia0 through /dev/nvidia254 */
+	for (unsigned nv_minor = 0; nv_minor < 255; nv_minor++) {
+		char nv_path[15] = { 0 };	// /dev/nvidiaXXX
+		sc_must_snprintf(nv_path, sizeof(nv_path), "/dev/nvidia%u",
+				 nv_minor);
 
-	udev_s->udev = udev_new();
-	if (udev_s->udev == NULL)
-		die("udev_new failed");
+		/* Stop trying to find devices after one is not found. In this manner,
+		 * we'll add /dev/nvidia0 and /dev/nvidia1 but stop trying to find
+		 * nvidia3 - nvidia254 if nvidia2 is not found. */
+		if (stat(nv_path, &sbuf) < 0) {
+			break;
+		}
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
 
-	udev_s->devices = udev_enumerate_new(udev_s->udev);
-	if (udev_s->devices == NULL)
-		die("udev_enumerate_new failed");
+	if (stat("/dev/nvidiactl", &sbuf) == 0) {
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
+	if (stat("/dev/nvidia-uvm", &sbuf) == 0) {
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
+	if (stat("/dev/nvidia-modeset", &sbuf) == 0) {
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
+}
 
-	if (udev_enumerate_add_match_tag(udev_s->devices, udev_s->tagname) < 0)
-		die("udev_enumerate_add_match_tag");
+/**
+ * Allow access to /dev/uhid.
+ *
+ * Currently /dev/uhid isn't represented in sysfs, so add it to the device
+ * cgroup if it exists and let AppArmor handle the mediation.
+ **/
+static void sc_udev_allow_uhid(int devices_allow_fd)
+{
+	struct stat sbuf;
 
-	if (udev_enumerate_scan_devices(udev_s->devices) < 0)
-		die("udev_enumerate_scan failed");
+	if (stat("/dev/uhid", &sbuf) == 0) {
+		dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			minor(sbuf.st_rdev));
+	}
+}
 
-	udev_s->assigned = udev_enumerate_get_list_entry(udev_s->devices);
-	if (udev_s->assigned == NULL)
-		rc = -1;
+/**
+ * Allow access to assigned devices.
+ *
+ * The snapd udev security backend uses udev rules to tag matching devices with
+ * tags corresponding to snap applications. Here we interrogate udev and allow
+ * access to all assigned devices.
+ **/
+static void sc_udev_allow_assigned(int devices_allow_fd, struct udev *udev,
+				   struct udev_list_entry *assigned)
+{
+	for (struct udev_list_entry * entry = assigned; entry != NULL;
+	     entry = udev_list_entry_get_next(entry)) {
+		const char *path = udev_list_entry_get_name(entry);
+		if (path == NULL) {
+			die("udev_list_entry_get_name failed");
+		}
+		struct udev_device *device =
+		    udev_device_new_from_syspath(udev, path);
+		if (device == NULL) {
+			die("cannot find device from syspath %s", path);
+		}
+		dev_t devnum = udev_device_get_devnum(device);
```

Comment:
As discussed on IRC, on my system a bunch of devices that do not have a device node are tagged right now, eg. video ports that apppear as separate devices under sys, but have no device node. The old code did a silly thing as well:
```
DEBUG: run_snappy_app_dev_add: /sys/devices/pci0000:00/0000:00:03.1/0000:23:00.0/0000:24:00.0/0000:25:00.0/drm/card0 snap_ohmygiraffe_ohmygiraffe
DEBUG: running snap-device-helper add snap_ohmygiraffe_ohmygiraffe /sys/devices/pci0000:00/0000:00:03.1/0000:23:00.0/0000:24:00.0/0000:25:00.0/drm/card0 226:0
DEBUG: run_snappy_app_dev_add: /sys/devices/pci0000:00/0000:00:03.1/0000:23:00.0/0000:24:00.0/0000:25:00.0/drm/card0/card0-DP-1 snap_ohmygiraffe_ohmygiraffe
>>>> see one below 
DEBUG: running snap-device-helper 
  add 
  snap_ohmygiraffe_ohmygiraffe 
  /sys/devices/pci0000:00/0000:00:03.1/0000:23:00.0/0000:24:00.0/0000:25:00.0/drm/card0/card0-DP-1
  0:0 <-- major:minor
...

DEBUG: running snap-device-helper add snap_ohmygiraffe_ohmygiraffe /sys/devices/pci0000:00/0000:00:03.1/0000:23:00.0/0000:24:00.0/0000:25:00.0/drm/renderD128 226:128
```

By assuming all devices have a devnum, we ended up adding `c 0:0 rwm` to allowed devices. This probably has no effect, since there's no block/char devices with major 0.

Reply:
The 0:0 really ended up in devices list:
```
$ sudo cat /sys/fs/cgroup/devices/snap.ohmygiraffe.ohmygiraffe/devices.list
[sudo] password for maciek: 
c 1:3 rwm
c 1:7 rwm
c 1:5 rwm
c 1:8 rwm
c 1:9 rwm
c 5:0 rwm
c 5:1 rwm
c 5:2 rwm
c 136:* rwm
c 137:* rwm
c 138:* rwm
c 139:* rwm
c 140:* rwm
c 141:* rwm
c 142:* rwm
c 143:* rwm
c 10:239 rwm
c 226:0 rwm
c 0:0 rwm   <---
c 226:128 rwm
```

Reply:
This is now fixed by 1353abe3bbe442860fe8837439996742d0d30398